fairscale/optim/oss.py

```diff
@@ -139,7 +133,16 @@ def partition_parameters(self) -> List[List[dict]]:
                     # Add this param to rank with smallest size.
                     rank = sizes.index(min(sizes))
                     param_lists[rank].append(param)
-                    sizes[rank] += param.numel()
+
+                    # We're partitioning the optimizer state,
```

Comment:
this is the real change, the partitioning was not taking into account the fact that the params are trainable or not, althgouh this is what counts for the optimizer state. The test case for Huggingface was kind of pathological for that, because there was one big non trainable parameter (goes to rank 0) and then a lot of cumulatively smaller trainable parameters, which all went to rank 1. This meant that the model was effectively optimized on rank 1, hence defying the whole sharding purpose